Yoast/wordpress-seo
# P1-1347 make yoast form use the new options framework

## Context


* Make Yoast Form use the new options services.

## Summary


This PR can be summarized in the following changelog entry:

* Fixes the Yoast setting forms by making use of the options services.

## Relevant technical choices:

* Added get/set options and network options route and actions (only options actions are used here)
* Added a `set_options` for bulk setting
* Added Form_Invalid_Exception to hold one or more field exceptions
* Network admin options are "registered" (put on allow list) for the AJAX action to check in the Options Initializer
* Applied network admin options AJAX solution to the Yoast form in the form of the Options_Form_Integration
* Added some missing options after rebasing (from trunk via feature branch). Note that the validators might not be correct yet, the goal is to save the data
* 

## Test instructions

### Test instructions for the acceptance test before the PR gets merged
This PR can be acceptance tested by following these steps:

#### Test the option forms
* Save without errors
* Reload the page and have the same data shown
* Cause validation errors (webmaster tools is easiest, insert `#!$#`) and verify invalid data did not save, but valid data did
* Ensure you have some custom post types and taxonomies active (you can use the Yoast Test helper for that)
* Bonus points for checking if the functionality is also correctly responding to the options
* Options to check on single and multisite:
  * General > Features
  * General > Integrations
  * General > Webmaster tools
  * General > First-time configuration
  * Search Appearance > General
  * Search Appearance > Content Types
    * Here the custom post types come into play 
  * Search Appearance > Media
  * Search Appearance > Taxonomies
    * Here the custom taxonomies come into play 
  * Search Appearance > Archives
  * Search Appearance > Breadcrumbs
  * Search Appearance > RSS
  * Social > Accounts
  * Social > Facebook
  * Social > Twitter
  * Social > Pinterest
* Options to check on multisite' network admin:
  * General > General
  * General > Features
    * If you disallow one, you should check a site under this multisite to verify they feature is now disabled
  * General > Integrations
    * If you disallow one, you should check a site under this multisite to verify they feature is now disabled

#### Test the multisite restore options
* In a multisite environment with at least 2 sites
* Change some options on site 1, including some privacy options that should not be copied over like the webmaster tools ones.
* Change some options on site 2 to ones differing from site 1 so you can check if it worked
* Go to multisite network admin options
* Save the inherit option under General > General to be site 1
* Go to General > Restore Site
* Restore the options of site 2
* Verify the options are as expected on site 2
  * Be sure to check none of the privacy options have values.

#### Test the option routes
* Install and active `JSON Basic Authentication` plugin to cheat the authentication a bit for easier testing (network activate)
* Fire some requests to the routes, we should have the following working:
  * GET yoast/v1/options -- retrieve all options
  * GET yoast/v1/options?options[]=facebook_site&options[]=twitter_site -- retrieve one or more options
  * POST yoast/v1/options -- with the data in the body: set one or more options
  * GET yoast/v1/network-admin-options -- retrieve all options
  * GET yoast/v1/network-admin-options?options[]=defaultblog&options[]=allow_content_analysis_active -- retrieve one or more options
  * POST yoast/v1/network-admin-options -- with the data in the body: set one or more options
* Example of curl POST request (that will produce failure), you should be able to figure out the other, right? I.e. change the domain to go to test single/multisite, change the last part of the URL to change network admin or normal options. Change the body to determine what to set. Remove the `-d "{}"` and change POST to GET for a GET request. Change the `admin:admin` part to be your user:password
```sh
curl -X POST --location "http://multisite.wordpress.test/wp-json/yoast/v1/network-admin-options" \
    -H "Content-Type: application/json" \
    -d "{
        	\"allow_content_analysis_active\": \"foo\",
        	\"allow_keyword_analysis_active\": \"bar\"
        }" \
    --basic --user admin:admin
```
* Do not forget to deactivate the `JSON Basic Authentication` plugin again after testing!

### Test instructions for QA when the code is in the RC




QA can test this PR by following these steps:

* Please test the feature branch as a whole instead of just this PR.

## Impact check

This PR affects the following parts of the plugin, which may require extra testing:

*

## UI changes


## Other environments


## Documentation


## Quality assurance


Fixes [P1-1347](https://yoast.atlassian.net/browse/P1-1347)
Fixes [P1-1104](https://yoast.atlassian.net/browse/P1-1104)
